rclcpp/include/rclcpp/rate.hpp

```diff
@@ -33,46 +34,59 @@ class RateBase
 
   virtual ~RateBase() {}
   virtual bool sleep() = 0;
-  virtual bool is_steady() const = 0;
+  virtual rcl_clock_type_t get_type() const = 0;
   virtual void reset() = 0;
 };
 
-using std::chrono::duration;
-using std::chrono::duration_cast;
-using std::chrono::nanoseconds;
-
 template<class Clock = std::chrono::high_resolution_clock>
-class GenericRate : public RateBase
+class [[deprecated("use rclcpp::Rate class instead of GenericRate")]] GenericRate
 {
 public:
   RCLCPP_SMART_PTR_DEFINITIONS(GenericRate)
 
-  explicit GenericRate(double rate)
-  : GenericRate<Clock>(
-      duration_cast<nanoseconds>(duration<double>(1.0 / rate)))
+  explicit GenericRate(double) {}
```

Comment:
The purpose to leave `GenericRate` class - is to warn developer that `Rate` should be used instead, from on now. All implementations were left empty to have only one one compiler deprecation warning message and thus to not cause a confusion of other inducted compiler errors (e.g. like "can not find such constructor or method"). It is supposed that developer will switch to `Rate` class after this message, so `GenericRate`'s implementations were emptied.
However, indeed, if warnings were disabled as errors on developer side, and he will refuse to follow the recommendations, this indeed might cause an undefined behavior.

> I'm not sure why this doesn't use `Rate` or `Rate` isn't implemented as `GenericRate` and we typedef a `Rate` from it or something if it doesn't make sense to have dual implementations

In vanilla version, `Rate` is a templated implementation of `GenericRate` (`using Rate = GenericRate<std::chrono::system_clock>;` or as you've mentioned, `typedef GenericRate<std::chrono::system_clock> Rate;` ). However, in current implementation, `Rate` concept now was changed. It is non-templated class, which can not be a template typedef of `GenericRate`. Moreover, if forcibly leave the implementation to use the `Rate` as `typedef` of `GenericRate` (as it now in vanilla), `GenericRate` as deprecated class, will cause compiler to warn when we will try to make the `Rate` as `typedef GenericRate<something>` .

Or did you mean to leave `GenericRate` untouched and have `Rate` as a completely new class, having dual-implementations? Wouldn't it cause even more confusion?

Finally, `GenericRate`'s implementation could be completely removed, which will cause additional compiler errors, but it never allow the developer to build their code with older API and thus to have an undefined behavior.

Reply:
I don't think that's the right philosophy to have on deprecation warnings. The warning is to let people know it will be depreciated soon in future releases, not that what you're currently using is already deprecated and will not work.

Reply:
> I don't think that's the right philosophy to have on deprecation warnings. The warning is to let people know it will be depreciated soon in future releases, not that what you're currently using is already deprecated and will not work.

Yes, agreed.  The old API needs to continue to work (with a compile-time deprecation warning).

Reply:
Okay, got it. I've restored the previous functionality of `GenericRate` class and API (including `is_steady()`) in the latest update